Velaptor/UI/TextBox.cs

```diff
@@ -106,23 +401,762 @@ public override void Render()
+                return;
+            }
+
+            this.inSelectionMode = false;
+            this.textSelection.Clear();
+        }
+    }
+
+    /// <summary>
+    /// Renders the text of the text box.
+    /// </summary>
+    private void RenderText()
+    {
+        if (this.font is null)
+        {
+            return;
+        }
+
+        var glyphsToRender = this.charBounds.Select((cb, i) =>
+        {
+            var inView = this.textAreaBounds.Contains(new Vector2(cb.bounds.Left, Position.Y)) &&
+                         this.textAreaBounds.Contains(new Vector2(cb.bounds.Right, Position.Y));
+
+            var glyphMetrics = this.font.Metrics.First(m => m.Glyph == cb.character);
+
+            var minIndex = Math.Min(this.selectionStartIndex, this.currentCharIndex);
+            var maxIndex = Math.Max(this.selectionStartIndex, this.currentCharIndex);
+
+            var lastCharSelected = this.textSelection.SelectionRect.Right >= this.charBounds.TextRight();
+
+            var inSelectedRange = lastCharSelected switch
+            {
+                true => this.inSelectionMode && i >= minIndex && i <= maxIndex,
+                false => this.inSelectionMode && i >= minIndex && i < maxIndex,
+            };
+
+            var clr = inSelectedRange
+                ? SelectedTextColor
+                : TextColor;
+
+            if (RenderAllText || inView)
+            {
+                return (glyphMetrics, clr);
+            }
+
+            glyphMetrics.GlyphBounds = glyphMetrics.GlyphBounds with { Width = 0 };
+
+            return (glyphMetrics, clr);
+        }).ToArray().AsSpan();
+
+        this.fontRenderer.Render(
+            this.font,
+            glyphsToRender,
+            (int)this.textPos.X,
+            (int)this.textPos.Y,
+            1f,
+            0f,
+            layer: 20);
+    }
+
+    /// <summary>
+    /// Adds a character to the text box.
+    /// </summary>
+    /// <param name="charKey">The character to add.</param>
+    private void Add(KeyCode charKey)
+    {
+        if (this.font is null)
+        {
+            return;
+        }
+
+        var character = charKey == KeyCode.Space ? " " : charKey.ToChar(this.currKeyState.AnyShiftKeysDown()).ToString();
+        var charIndexAtEnd = this.currentCharIndex >= this.text.LastCharIndex();
+        var cursorAtEnd = this.preTextBoxState.CursorAtEnd;
+
+        var insertIndex = charIndexAtEnd && cursorAtEnd
+            ? this.text.Length
+            : this.currentCharIndex;
+
+        var prevHeight = this.textSize.Height;
+
+        this.text.Insert(insertIndex, character);
+        this.textSize = this.font.Measure(this.text.ToString());
+
+        UpdateBounds(charKey); // NEW
+        CalcYPositions(prevHeight, this.textSize.Height);
+
+        this.currentCharIndex = charIndexAtEnd
+            ? this.text.LastCharIndex()
+            : this.currentCharIndex + 1;
+    }
+
+    /// <summary>
+    /// Removes the a character from the text box.
```

Comment:
Grammar - If you're talking about a specific char, it's best to set it apart so the sentence flows.  You can also use "a character" if that's the actual name. 
```suggestion
    /// Removes the 'a' character from the text box.
```